float32 px
float32 py
float32 ow
int32 mode